src/main/java/edu/kit/kastel/sdq/intelligrade/extensions/guis/table/AnnotationsTreeNode.java

```diff
@@ -0,0 +1,311 @@
+                    .orElse("");
+            default -> throw new IllegalStateException("No table data at column %d".formatted(column));
+        };
+    }
+
+    /**
+     * Creates a list over the children of this node.
+     *
+     * @return a {@link List} over the children of this node
+     */
+    public List<AnnotationsTreeNode> listChildren() {
+        return Collections.list(children()).stream()
+                .map(AnnotationsTreeNode.class::cast)
+                .toList();
+    }
+
+    public abstract Object getValueAt(int column);
+
+    public abstract List<Annotation> listAnnotations();
+
+    public boolean removeIf(Predicate<Annotation> shouldRemove) {
+        boolean hasAnyChanged = false;
+
+        for (var child : listChildren()) {
+            boolean hasChanged = child.removeIf(shouldRemove);
+            if (hasChanged) {
+                hasAnyChanged = true;
+            }
+
+            // remove the child if it should be removed
+            if (child instanceof AnnotationNode annotationNode) {
+                if (shouldRemove.test(annotationNode.getAnnotation())) {
+                    remove(child);
+                    hasAnyChanged = true;
+                }
+                continue;
+            }
+
+            // if a group has only one child, replace the group with the child
+            if (hasChanged && child.getChildCount() == 1) {
+                var onlyChild = child.listChildren().getFirst();
+                replace(child, onlyChild);
+                continue;
+            }
+
+            // remove empty children (e.g. groups which have no annotations)
+            //
+            // Note: if annotations are removed one by one, a group will never be empty,
+            //       because of the above code, but if multiple annotations are removed at once,
+            //       a group could become empty.
+            if (hasChanged && child.getAllowsChildren() && child.getChildCount() == 0) {
+                remove(child);
+            }
+        }
+
+        return hasAnyChanged;
+    }
+
+    public void replace(AnnotationsTreeNode oldNode, AnnotationsTreeNode newNode) {
+        int index = getIndex(oldNode);
+        if (index == -1) {
+            throw new IllegalArgumentException("oldNode is not a child of this node");
+        }
+
+        remove(index);
+        insert(newNode, index);
+    }
+
+    /**
+     * Sorts all children of this node recursively.
+     *
+     * @param comparator the comparator that defines the sorting
+     */
+    public void sort(Comparator<? super AnnotationsTreeNode> comparator) {
+        List<AnnotationsTreeNode> children = new ArrayList<>(listChildren());
+        this.removeAllChildren();
+
+        for (var child : children) {
+            child.sort(comparator);
+        }
+
+        children.sort(comparator);
+
+        children.forEach(this::add);
+    }
+
+    public static class AnnotationNode extends AnnotationsTreeNode {
```

Comment:
~I think a record would work here, since you are using trivial equals and hashCode methods~

Nvmd, you need inheritance